components/sup/src/manager/service/config.rs

```diff
@@ -170,8 +215,54 @@ impl Cfg {
         Ok(())
     }
 
-    fn load_environment(&mut self, package: &Pkg) -> Result<()> {
-        let var_name = format!("{}_{}", ENV_VAR_PREFIX, package.name)
+    fn open_user_toml<P: PackageForConfig>(&mut self, package: &P) -> Option<(File, PathBuf)> {
+        let recommended_path = package.user_config_path().join("user.toml");
+        let ucps = if let Some(ref ucp) = self.user_config_path {
+            // Marking the path as not deprecated, even if the user
+            // config path points to the old path. If this is the
+            // case, we already warned about it.
+            vec![UserConfigPath::Recommended(ucp.get_path().clone())]
+        } else {
+            vec![
+                UserConfigPath::Recommended(recommended_path.clone()),
+                UserConfigPath::Deprecated(package.svc_path().join("user.toml")),
+            ]
+        };
+        for ucp in ucps {
+            match File::open(ucp.get_path()) {
+                Ok(file) => {
+                    if let UserConfigPath::Deprecated(_) = ucp {
+                        outputln!(
+                            "The {} at {} is deprecated, consider putting it in {}",
+                            "user configuration location",
+                            ucp.get_path().display(),
+                            recommended_path.display(),
+                        );
+                    }
+                    if self.user_config_path.is_none() {
+                        self.user_config_path = Some(ucp.clone());
+                    }
+                    return Some((file, ucp.steal_path()));
+                }
+                Err(e) => {
+                    debug!(
+                        "Failed to open 'user.toml', {}, {}",
+                        ucp.get_path().display(),
+                        e,
+                    );
+                }
+            }
+        }
+        // No user config found at all. Next time try loading only the
```

Comment:
When is "next time"? As far as I can see, this is only ever called when creating a new `Cfg` struct.

I presume this is laying the groundwork for refreshing the `user.toml` content based on file watch events... in that case, is the intention to never change which `user.toml` path is being looked at? If this starts up and a `user.toml` in the new location is not present, but one at the old path is present, won't this always look at the old path, even if a new one shows up later? If that's the case, it may simplify this code and make it easier to reason about if we split this function into two functions: one that does the initial determination of which path we're going to look at, and the other that does the actual consumption of content. Actually, in that case, we could probably just use the original code after we figured out which path to consult.

Reply:
Yeah, we basically want to stick to a single path, which is determined at the time when we load the config for the first time.

I'll have a look at splitting the functions. I also had an idea of cleaning up the load_{default,user,environment} functions, so they don't take the `self` parameter. This would be to avoid using `Cfg::default`, which forces the `user_config_parameter` to be an `Option`, because it's value is set sometime later in the `new` function.